packages/datadog-instrumentations/src/kafkajs.js

```diff
@@ -147,35 +155,63 @@ addHook({ name: 'kafkajs', file: 'src/index.js', versions: ['>=1.4'] }, (BaseKaf
   return Kafka
 })
 
-const wrapFunction = (fn, startCh, finishCh, errorCh, extractArgs) => {
+const wrapFunction = (fn, startCh, finishCh, errorCh, extractArgs, kafkaClusterIdPromise) => {
   return typeof fn === 'function'
     ? function (...args) {
-      const innerAsyncResource = new AsyncResource('bound-anonymous-fn')
-      return innerAsyncResource.runInAsyncScope(() => {
-        const extractedArgs = extractArgs(args)
-        startCh.publish(extractedArgs)
-        try {
-          const result = fn.apply(this, args)
-          if (result && typeof result.then === 'function') {
-            result.then(
-              innerAsyncResource.bind(() => finishCh.publish(undefined)),
-              innerAsyncResource.bind(err => {
-                if (err) {
-                  errorCh.publish(err)
-                }
-                finishCh.publish(undefined)
-              })
-            )
-          } else {
+      kafkaClusterIdPromise.then((clusterId) => {
+        const innerAsyncResource = new AsyncResource('bound-anonymous-fn')
+        return innerAsyncResource.runInAsyncScope(() => {
+          const extractedArgs = extractArgs(args, clusterId)
+
+          startCh.publish(extractedArgs)
+          try {
+            const result = fn.apply(this, args)
+            if (result && typeof result.then === 'function') {
+              result.then(
+                innerAsyncResource.bind(() => finishCh.publish(undefined)),
+                innerAsyncResource.bind(err => {
+                  if (err) {
+                    errorCh.publish(err)
+                  }
+                  finishCh.publish(undefined)
+                })
+              )
+            } else {
+              finishCh.publish(undefined)
+            }
+            return result
+          } catch (e) {
+            errorCh.publish(e)
             finishCh.publish(undefined)
+            throw e
           }
-          return result
-        } catch (e) {
-          errorCh.publish(e)
-          finishCh.publish(undefined)
-          throw e
-        }
+        })
       })
     }
     : fn
 }
+
+const getKafkaClusterId = async (kafka) => {
+  if (kafka._ddKafkaClusterId) {
+    return kafka._ddKafkaClusterId
+  }
+
+  if (!kafka.admin) {
+    return null
+  }
+
+  const admin = kafka.admin()
+
+  if (!admin.describeCluster) {
+    return null
+  }
+
+  await admin.connect()
+  const clusterInfo = await admin.describeCluster()
+  const clusterId = clusterInfo?.clusterId
+  await admin.disconnect()
```

Comment:
You can actually disconnect as soon as you've gotten `clusterInfo`, _and_ you don't have to `await` it, since you don't _need_ it to be disconnected to proceed.

That said, all `async/await` syntax in this file should be converted to promise chains, for performance reasons.

Reply:
Done, converted this to use promise chains and the admin disconnects as soon as we've got `clusterInfo`